equinox-shipment/Domain/FinalizationTransactionProcessor.fs

```diff
@@ -0,0 +1,59 @@
+module FinalizationTransactionProcessor
+
+    type Service(transactions: FinalizationTransaction.Service, containers: Container.Service, shipments: Shipment.Service) =
+
+        // Transaction ID can be a md5 hash of all shipments ID
+        let execute (transactionId: string) : FinalizationTransaction.Events.Event option -> Async<bool> =
+            let rec loop (update: FinalizationTransaction.Events.Event option) =
+                async {
+                    let loop event = loop (Some event)
+
+                    let! action =
+                        transactions.Apply(transactionId, update)
+
+                    match action with
+                    | FinalizationTransaction.Action.AssignShipments (containerId, shipmentIds) ->
+                        let! result =
+                            shipmentIds
+                            |> Array.map (fun sId ->
+                                async {
+                                    let! res = shipments.Execute(sId, Shipment.Command.Assign containerId)
+                                    return (sId, res)
+                                }
+                            )
+                            |> Async.Parallel
+
+                        let failures =
+                            Array.partition snd result
+                            |> snd
+                            |> Array.map fst
+
+                        return!
+                            if not <| Array.isEmpty failures then
+                                loop (FinalizationTransaction.Events.RevertRequested (containerId, failures))
+                            else
+                                loop (FinalizationTransaction.Events.AssignmentCompleted (containerId, shipmentIds))
+
+                    | FinalizationTransaction.Action.FinalizeContainer (containerId, shipmentIds) ->
+                        do! containers.Execute (containerId, Container.Command.Finalize shipmentIds)
+
+                        return! loop (FinalizationTransaction.Events.FinalizationCompleted containerId)
+
+                    | FinalizationTransaction.Action.RevertAssignment (containerId, shipmentIds) ->
+                        do! shipmentIds
```

Comment:
let! _ = Async.Parallel(seq { for sId in shipmentIds -> shipments.Execute... } )
need to throttle as above

Reply:
Updated thanks!